src/QMCApp/QMCMain.h

```diff
@@ -33,19 +33,34 @@ namespace qmcplusplus
  * This is a generalized QMC application which can handle multiple ParticleSet,
  * TrialWaveFunction and QMCHamiltonian objects.
  */
-class QMCMain : public QMCMainState, public QMCAppBase
+class QMCMain : public MPIObjectBase, public QMCAppBase
 {
 public:
-  ///constructor
   QMCMain(Communicate* c);
 
-  ///destructor
-  ~QMCMain() override;
+  ~QMCMain();
 
   bool validateXML() override;
   bool execute() override;
 
+  ParticleSetPool& getParticlePool() { return *particle_set_pool_; }
+
 private:
+  /// ParticleSet Pool
+  std::unique_ptr<ParticleSetPool> particle_set_pool_;
+
+  /// TrialWaveFunction Pool
+  std::unique_ptr<WaveFunctionPool> psi_pool_;
+
+  /// QMCHamiltonian Pool
+  std::unique_ptr<HamiltonianPool> ham_pool_;
+
+  /// current MCWalkerConfiguration
+  MCWalkerConfiguration* qmc_system_;
+
+  ///Global estimators defined outside of <qmc> nodes
+  std::optional<EstimatorManagerInput> estimator_manager_input_;
+
```

Comment:
The MainState was introduced to improve the organization of member objects of QMCMain.
The down side is it hurt readability and having QMCMain as a base is awkward.
Can we have an anonymous struct here?
```
// QMCMainState
struct {
    /// ParticleSet Pool
  std::unique_ptr<ParticleSetPool> particle_set_pool_;
 ...
};
```
Maybe we can have a bit more organization this way.

Reply:
Cannot use anonymous struct.
```
   51 |   std::unique_ptr<ParticleSetPool> particle_set_pool_;
      |                                    ^~~~~~~~~~~~~~~~~~
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:51:36: error: member ‘std::unique_ptr<qmcplusplus::ParticleSetPool> qmcplusplus::QMCMain::<unnamed struct>::particle_set_pool_’ with destructor not allowed in anonymous aggregate
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:51:36: error: member ‘std::unique_ptr<qmcplusplus::ParticleSetPool> qmcplusplus::QMCMain::<unnamed struct>::particle_set_pool_’ with copy assignment operator not allowed in anonymous aggregate
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:54:37: error: member ‘std::unique_ptr<qmcplusplus::WaveFunctionPool> qmcplusplus::QMCMain::<unnamed struct>::psi_pool_’ with constructor not allowed in anonymous aggregate
```
Then ignore my suggestion.